app/models/app.rb

```diff
@@ -88,8 +88,21 @@ def ready?
     integrations.ready? and config&.ready?
   end
 
+  # NOTE: fetches and uses latest build numbers from the stores, if added,
+  # to reduce build upload rejection probability
   def bump_build_number!
     with_lock do
+      ios_latest_build_number = ios_store_provider&.latest_build_number
+      android_latest_build_number = android_store_provider&.latest_build_number
+
+      if ios_latest_build_number.present? && build_number < ios_latest_build_number
+        self.build_number = ios_latest_build_number
+      end
+
+      if android_latest_build_number.present? && build_number < android_latest_build_number
+        self.build_number = android_latest_build_number
+      end
+
       self.build_number = build_number.succ
```

Comment:
You can try a neater (but potentially clever) way for this:

```suggestion
    self.build_number = [
      ios_store_provider&.latest_build_number, 
      android_store_provider&.latest_build_number, 
      build_number.succ
    ].compact.max
```

Reply:
hm, need to `succ` the max of all, but yeah I can try this.